sklearn/calibration.py

```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
+        method_name = "predict_proba"
+        method = getattr(clf_fitted, "predict_proba")
+    else:
+        raise RuntimeError("'base_estimator' has no 'decision_function' or "
+                           "'predict_proba' method.")
+    if return_string:
+        return method_name
+    else:
+        return method
 
-    classes : array-like, shape (n_classes,), optional
-            Contains unique classes used to fit the base estimator.
-            if None, then classes is extracted from the given target values
-            in fit().
 
-    See also
-    --------
-    CalibratedClassifierCV
+def _reshape_preds(preds, method, n_classes):
+    """Reshape predictions when classification binary.
 
-    References
+    Parameters
     ----------
-    .. [1] Obtaining calibrated probability estimates from decision trees
-           and naive Bayesian classifiers, B. Zadrozny & C. Elkan, ICML 2001
+    preds : array-like
+        Predictions.
 
-    .. [2] Transforming Classifier Scores into Accurate Multiclass
-           Probability Estimates, B. Zadrozny & C. Elkan, (KDD 2002)
+    method : {'decision_function', 'predict_proba'}
+        Method used to obtain the predictions.
 
-    .. [3] Probabilistic Outputs for Support Vector Machines and Comparisons to
-           Regularized Likelihood Methods, J. Platt, (1999)
+    n_classes : int
+        Number of classes.
 
-    .. [4] Predicting Good Probabilities with Supervised Learning,
-           A. Niculescu-Mizil & R. Caruana, ICML 2005
+    Returns
+    -------
+    preds : array-like, shape (n_samples, 1)
+        Reshaped predictions
     """
-    @_deprecate_positional_args
-    def __init__(self, base_estimator, *, method='sigmoid', classes=None):
-        self.base_estimator = base_estimator
-        self.method = method
-        self.classes = classes
-
-    def _preproc(self, X):
-        n_classes = len(self.classes_)
-        if hasattr(self.base_estimator, "decision_function"):
-            df = self.base_estimator.decision_function(X)
-            if df.ndim == 1:
-                df = df[:, np.newaxis]
-        elif hasattr(self.base_estimator, "predict_proba"):
-            df = self.base_estimator.predict_proba(X)
-            if n_classes == 2:
-                df = df[:, 1:]
-        else:
-            raise RuntimeError('classifier has no decision_function or '
-                               'predict_proba method.')
+    if method == 'decision_function':
+        if preds.ndim == 1:
+            preds = preds[:, np.newaxis]
+    elif method == 'predict_proba':
+        if n_classes == 2:
+            preds = preds[:, 1:]
+    else:
+        raise RuntimeError("'method' needs to be one of 'decision_function' "
+                           "or 'predict_proba'.")
+    return preds
 
-        idx_pos_class = self.label_encoder_.\
-            transform(self.base_estimator.classes_)
 
-        return df, idx_pos_class
+def _get_predictions(pred_method, X, label_encoder_):
```

Comment:
Here and in other places:

- since it's not an attribute in this context, we should just name it `label_encoder_`
- however, since we only use its `classes_` attribute, we should just pass `classes_`. Or even better, `n_classes`. It makes it more explicit that what this function ultimately needs to know is the number of classes. How we get it should be an implementation detail.

Reply:
Good point. I've amended for `_get_predictions`.

For `_fit_calibrator` and for `_CalibratedClassiferPipeline.predict_proba()` we use `label_encoder_.transform()` to get the indices of the classes present. I only input `label_encoder_.classes_` and instantiate a new `LabelEncoder` instance but I think passing `label_encoder_` is better? WDYT?